### ISSUE TITLE
Random/intermittent "Input data ends in the middle of a stack." error on Windows with softbuffer

### ISSUE THREAD

**Issue by LoganDark** [flamegraph-rs/flamegraph#207]
I am trying to profile a [`softbuffer`](https://github.com/john01dav/softbuffer) application, because I have to implement my own color blending (among other things). I'm using `flamegraph` directly like so:

```
PS Microsoft.PowerShell.Core\FileSystem::\\wsl$\Ubuntu-20.04-WSL2\home\logandark> flamegraph -- .\.cargo\target\x86_64-pc-windows-gnu\release\application.exe
Error: unable to collapse generated profile data

Caused by:
    Input data ends in the middle of a stack.
```

due to `cargo` being unable to run a normal build in Windows-land because of the WSL filesystem.

If I go through the work (and look up the command -_-) to set up a `CARGO_TARGET_DIR` to run `cargo flamegraph` in Windows-land, it completes successfully:

```
PS Microsoft.PowerShell.Core\FileSystem::\\wsl$\Ubuntu-20.04-WSL2\home\logandark\CLionProjects\application> New-Item -Name 'CARGO_TARGET_DIR' -Value 'C:\Users\LoganDark\.cargo\target' -ItemType Variable -Path Env: -Force

Name                           Value
----                           -----
CARGO_TARGET_DIR               C:\Users\LoganDark\.cargo\target


PS Microsoft.PowerShell.Core\FileSystem::\\wsl$\Ubuntu-20.04-WSL2\home\logandark\CLionProjects\application> cargo flamegraph
WARNING: profiling without debuginfo. Enable symbol information by adding the following lines to Cargo.toml:

[profile.release]
debug = true

Or set this environment variable:

CARGO_PROFILE_RELEASE_DEBUG=true

writing flamegraph to "flamegraph.svg"
```

I'm like 99% sure this is because the exe I was trying to use before was cross-compiled by mingw on WSL.

**Comment by LoganDark** [comment]
OK, changed the title because I'm now getting this with native `cargo flamegraph`. Not sure why.

**Comment by djc** [comment]
Are you using the dtrace backend or the blondie backend?

cc @nico-abram

**Comment by LoganDark** [comment]
![image](https://user-images.githubusercontent.com/4723091/173503450-12aafebd-d1f4-4990-b115-d36e99840293.png)

Ok this is... weird. And wrong.

In the rare case it actually works, there are random win32/winit stacks appearing in the middle of stuff like `stackblur` that has absolutely nothing to do with any of that stuff.

Maybe this is because the stacks actually ARE being completely misinterpreted/corrupted by blondie

**Comment by nico-abram** [comment]
@LoganDark I tried building the stackblur `fruit` example on windows (msvc target) and it seems to work alright (10 out of 10 times it finished succesfully and the flamegraph looks ok).

Is your application open source, could I try building it from somewhere?

Also, I'm a bit confused by the cross compile step: Are you just running something like `cargo build --target x86_64-pc-windows-gnu` from within WSL, and then pointing flamegraph to the WSL filesystem in `\\wsl$\...`?

**Comment by LoganDark** [comment]
> Is your application open source, could I try building it from somewhere?

I'm just running `winit` with `softbuffer`, any app with these 2 libs has the issue.

> Also, I'm a bit confused by the cross compile step: Are you just running something like `cargo build --target x86_64-pc-windows-gnu` from within WSL, and then pointing flamegraph to the WSL filesystem in `\\wsl$\...`?

Yep

**Comment by Rodrigodd** [comment]
I am also being hit by this issue in one of my projects. I can reproduce this by just running a example of the [winit](https://github.com/rust-windowing/winit) repo, in the last master commit ([37c0f61](https://github.com/rust-windowing/winit/commit/37c0f615cf0cc6817fd7d7a70c36b6d5c0fb4f2c)).

Running like this (I remove the `println!("{event:?}");` from the example, if this makes any diference):

```
D:\repos\rust\winit>cargo flamegraph --example window
```

Running the command above 15 times (just running the command and closing the window immediately, but the time running appears to make no diference),  have give me the following output 5 times:

```
D:\repos\rust\winit>cargo flamegraph --example window
    Finished release [optimized] target(s) in 0.17s

WARNING: profiling without debuginfo. Enable symbol information by adding the following lines to Cargo.toml:

[profile.release]
debug = true

Or set this environment variable:

CARGO_PROFILE_RELEASE_DEBUG=true

Error: unable to collapse generated profile data

Caused by:
    Input data ends in the middle of a stack.
```

Can also confirm that many of the frames have a branch starting with a ``ntdll`KiUserCallbackDispatcher``, (like in ``window`core::fmt::Formatter::debug_struct_field2_finish`` if you restore the `println!("{event:?}");` in the example). I also profile the same example using Windows Performance Recorder and Windows Performance Analyzer, but I could not find these branchs.

Testing some random cli programs that don't use `winit` didn't reproduced neither problems.

Tested with flamegraph 0.6.2, on Windows 10.0.19044, WPA 10.0.14393.0, and Rust version 1.66.0 with target `stable-x86_64-pc-windows-msvc`.

**Comment by nico-abram** [comment]
Thanks for the report @Rodrigodd 

I tried reproducing the issue by running the winit example 20 times, but it always generated the flamegraph for me. Something is probably failing but I'm not certain what. My first guess was that something else in your system was acquiring the kernel logging session (Only one process can use at the same time as I understand it) and I'm not handling that properly. After running the example 3 times and starting an xperf  "NT Kernel Logger" trace session while it is running (Should be the same mechanism as Windows Performance Recorder), I managed to trigger an error thanks to that, but the error message was different:

```
could not find dtrace and could not profile using blondie: Other(WIN32_ERROR(4201), "The instance name passed was not recognized as valid by a WMI data provider.\r\n", "ControlTraceA STOP ProcessTrace")
```

So I'm not entirely sure what's going on there. Did you pherhaps profile with WPR and flamegraph at the same time?

I tried to find the issue with `KiUserCallbackDispatcher`. Is this it?
![imagen](https://user-images.githubusercontent.com/24706838/217978083-c0c4700c-9c53-4534-8119-4f9850c292fd.png)

This does certainly look odd. `KiUserCallbackDispatcher` seems to be a trampoline to call back into user code from kernel code, but the stack below this has no syscall that would have switched to kernel code from what I can see.

I'm using microsoft's dbghelp.dll to resolve the call stacks. I'm working on a new version of the library that uses [pdb-addr2line](https://crates.io/crates/pdb-addr2line) and [symsrv](https://crates.io/crates/symsrv) instead which generates the following flamegraph:

![imagen](https://user-images.githubusercontent.com/24706838/217978531-18efa7ba-64ce-42cd-8766-12efc64667ee.png)

My original motivation was avoiding the dependency on the dll, and that dbghelp can be somewhat slow, but it seems either thanks to the library change or as a side effect of my changes, that bug is fixed. If you'd like to test it, just build flamegraph after changing the blondie dependency from "0.3" to "0.4".

**Comment by Rodrigodd** [comment]
@nico-abram Can confirm that updating the version of blondie fixed the problem.

Testing with flamegraph master ([ee821ba](https://github.com/flamegraph-rs/flamegraph/commit/ee821bab17fa0cf296e01cd829bd44d2b5f0c55c)) still reproduces the problem, but updating blondie to "0.4.0" fixed them.

Can also confirm that the frames look much more clear now, no `KiUserCallbackDispatcher` or ``VCRUNTIME140`memcpy`` calls anymore.

Only issue that I found is that the first two times I run the flamegraph command it has freeze for more than 2 minutes or so (I canceled it at that point), with 0% CPU utilization. But subsequent calls run smoothly, and I could not reproduce the problem anymore.

**Comment by nico-abram** [comment]
That's great!

It was most likely downloading symbols from microsoft's symbol servers for system dlls. Maybe I should disable that by default and only enable it if the environment variable (_NT_SYMBOL_PATH) is set.

**Comment by Rodrigodd** [comment]
> It was most likely downloading symbols from microsoft's symbol servers for system dlls when it froze. Maybe I should disable that by default and only enable it if the environment variable (_NT_SYMBOL_PATH) is set.

@nico-abram Oh, so it is okay. I have that env var already setup, so I am okay either way.

**Comment by Bjohnson131** [comment]
I'm seeing this issue with msvc on flamegraph 0.6.2, with 1.70.0-nightly running the mage core examples: https://github.com/baad-c0de/mage-core

**Comment by nico-abram** [comment]
@Bjohnson131 I don't think 6.2 includes the change that closed this. Could you check if cargo-flamegraph manually built from the latest commit fixes the problem?

**Comment by Bjohnson131** [comment]
Checking.. will report back.

**Comment by Bjohnson131** [comment]
@nico-abram confirmed that installing flamegraph via the following  command fixes the issue on windows: 

`cargo install --git https://github.com/flamegraph-rs/flamegraph --branch main`

**Comment by djc** [comment]
I'll see if I can get a release out soon.

**Comment by djc** [comment]
Published 0.6.3.
